src/rpc/network/socket_communicator.h

```diff
@@ -196,28 +210,41 @@ class SocketReceiver : public Receiver {
   /*!
    * \brief socket for each client connections
    */ 
-  std::unordered_map<int /* Sender (virutal) ID */, std::shared_ptr<TCPSocket>> sockets_;
+  std::vector<std::unordered_map<int /* Sender (virutal) ID */,
+    std::shared_ptr<TCPSocket>>> sockets_;
 
   /*!
    * \brief Message queue for each socket connection
    */ 
-  std::unordered_map<int /* Sender (virtual) ID */, std::shared_ptr<MessageQueue>> msg_queue_;
+  std::unordered_map<int /* Sender (virtual) ID */,
+    std::shared_ptr<MessageQueue>> msg_queue_;
```

Comment:
it's a bit strange. you made msg_queue here is per socket, but the message queue in the sender is per thread.

Reply:
Good question. Let me try to unify them. Maybe we can get rid of the semaphore.

Reply:
Just took a closer look. Seems like I did this to keep RecvFrom(msg, send_id) easy. If multiple senders share one msg_queue, this RecvFrom will be hard to write. So I think we can keep current design.